// ng-print/esm2015/lib/ng-print.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Inject } from "@angular/core";
import { DOCUMENT } from '@angular/common';
import { WINDOW } from './window.ref';
export class NgPrintDirective {
    /**
     * @param {?} document
     * @param {?} window
     */
    constructor(document, window) {
        this.document = document;
        this.window = window;
    }
    /**
     * @return {?}
     */
    print() {
        /** @type {?} */
        const x = this.document.getElementsByTagName("head")[0].innerHTML;
        /** @type {?} */
        const iframe = this.document.createElement("iframe");
        iframe.style.display = 'none';
        this.document.body.appendChild(iframe);
        iframe.contentWindow.document.write(x);
        iframe.contentWindow.document.write(this.elementToPrint.innerHTML);
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.window.frames[0].print();
        }), 500);
        // remove iframe after print or cancelation
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            var a = this.document.querySelectorAll('iframe');
            a[0].parentNode.removeChild(a[0]);
        }), 500);
    }
}
NgPrintDirective.decorators = [
    { type: Directive, args: [{
                selector: "button[ngPrint]"
            },] }
];
/** @nocollapse */
NgPrintDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WINDOW,] }] }
];
NgPrintDirective.propDecorators = {
    elementToPrint: [{ type: Input }],
    print: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    NgPrintDirective.prototype.elementToPrint;
    /**
     * @type {?}
     * @private
     */
    NgPrintDirective.prototype.document;
    /**
     * @type {?}
     * @private
     */
    NgPrintDirective.prototype.window;
}
