// ng-print/esm5/lib/ng-print.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Inject } from "@angular/core";
import { DOCUMENT } from '@angular/common';
import { WINDOW } from './window.ref';
var NgPrintDirective = /** @class */ (function () {
    function NgPrintDirective(document, window) {
        this.document = document;
        this.window = window;
    }
    /**
     * @return {?}
     */
    NgPrintDirective.prototype.print = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var x = this.document.getElementsByTagName("head")[0].innerHTML;
        /** @type {?} */
        var iframe = this.document.createElement("iframe");
        iframe.style.display = 'none';
        this.document.body.appendChild(iframe);
        iframe.contentWindow.document.write(x);
        iframe.contentWindow.document.write(this.elementToPrint.innerHTML);
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.window.frames[0].print();
        }), 500);
        // remove iframe after print or cancelation
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var a = _this.document.querySelectorAll('iframe');
            a[0].parentNode.removeChild(a[0]);
        }), 500);
    };
    NgPrintDirective.decorators = [
        { type: Directive, args: [{
                    selector: "button[ngPrint]"
                },] }
    ];
    /** @nocollapse */
    NgPrintDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WINDOW,] }] }
    ]; };
    NgPrintDirective.propDecorators = {
        elementToPrint: [{ type: Input }],
        print: [{ type: HostListener, args: ['click',] }]
    };
    return NgPrintDirective;
}());
export { NgPrintDirective };
if (false) {
    /** @type {?} */
    NgPrintDirective.prototype.elementToPrint;
    /**
     * @type {?}
     * @private
     */
    NgPrintDirective.prototype.document;
    /**
     * @type {?}
     * @private
     */
    NgPrintDirective.prototype.window;
}
